### ISSUE TITLE
Spacemacs users seem to struggle with debugging

### ISSUE THREAD

**Issue by bradyt** [syl20bnr/spacemacs#12667]
When issues arise on public projects, or IRC, or Gitter, Spacemacs users seem unable to apply basic troubleshooting methods that apply to Emacs.

I have watched the Gitter channel, hoping to see some sort of method, but have not seen such.

We encounter such users on Freenode's IRC frequently. Users report that they cannot find as much help on Spacemacs's Gitter channel.

What I am referring to, is strategies like checking emacs -Q, or emacs -q -f package-initialize, which both lead pretty smoothly into bisecting an init file by commenting out the lower half and repeating.

When offering such methods to Spacemacs users, they seem to be at a loss. For one thing, there is a huge difference between emacs -q, and running Spacemacs with zero configuration.

Is it possible for the Spacemacs project to provide strategies nearly as methodical? I don't know if a Spacemacs user can comment the bottom half of the config file and recurse. Can you provide an analogue in your resources? I don't know how to go smoothly from `emacs -q` to bisecting a Spacemacs config.

There is also the matter that they don't seem equipped on how to present problems at other projects. If you ask a new Spacemacs user to present their issue as a minimal reproduce, they seem unable to create such a recipe. Can you have some resources for them on how to go from a perceived bug, to reporting a minimal broken example to other Emacs projects?

Is there a solution for this? How much of this distance between communities is permanent? Can you provide tutorials, or new strategies, that parallel the straightforward approaches in Emacs, described above?

The support strategies described for Emacs, go a long way towards giving users tools to solve problems themselves, saving a lot of time for everyone. I realize that you might consider a preconfigured Emacs a time saver, but that is an upfront savings. Perhaps you can save them time down the road as well, by providing troubleshooting strategies that parallel those available to Emacs users. And as I said, they often end up not finding the help they need on your Gitter channel, and looking for help at the Freenode channel for Emacs. Much frustration ensues on both sides.

**Comment by smile13241324** [comment]
There is a troubleshooting section in DOCUMENTATION.org which contains some basic hints for troubleshooting.
```org
* Troubleshoot
** Loading fails
If any errors happen during the loading the mode-line will turn red and the
errors should appear inline in the startup buffer. Spacemacs should still be
usable; if it is not then restart Emacs with =emacs --debug-init= and open a
[[https://github.com/syl20bnr/spacemacs/issues][Github issue]] with the backtrace.

** Upgrading/Downgrading Emacs version
To ensure that packages are correctly compiled for the new Emacs version you
installed, be sure to run the interactive command =spacemacs/recompile-elpa=
with ~SPC u SPC SPC spacemacs/recompile-elpa~.
```
Also there are some requirements in CONTRIBUTING.org
```org
* Reporting issues
Issues have to be reported on our [[https://github.com/syl20bnr/spacemacs/issues][issues tracker]]. Please:
- Check that the issue has not already been reported.
  - This can be achieved by searching keywords on the [[https://github.com/syl20bnr/spacemacs/issues][issues tracker]].
- Check that the issue has not been fixed in the =develop= version of Spacemacs.
  - This can be achieved by running Spacemacs on the =develop= branch and trying
    to reproduce the bug here. You can also check at the [[https://github.com/syl20bnr/spacemacs/tree/develop][source code]] to see if
    it has been changed/corrected.
- Try to use a clear title, and describe your problem with complete sentences.
  See also [[https://github.com/syl20bnr/spacemacs/wiki/Debugging#how-to-make-a-great-bug-report][How to make a great bug report]] in the wiki.
- Include the following information in your issue:
  - The output of =SPC h d s= (=M-m h d s= in Emacs style), which gives the
    versions information about your installation.
  - If relevant, include the mode in which the problem arise (e.g. javascript
    files, =org-mode=, etc…).
  - If possible, try to include details on how to reproduce it, like a step by
    step guide.
```

In addition there are some standard reproduction strategies recommended to issue authors by the maintainers like reproducing an issue with a default config. Though this is not part of DOCUMENTATION.org right now.

I will add some additional instructions to help new users. Nevertheless we are not responsible for spacemacs users complaining on the emacs chat either. Spacemacs users should also get familiar with elisp and the vim bindings, at least when they are using evil, but this has no top priority and may only be necessary later when more fancy user-configs are required.

**Comment by bradyt** [comment]
TL;DR: If I could point users to Spacemacs's documentation on how to comment out half of their init file and recurse, that would be an awesome start. So if you had a page or a (linkable) section title with, "How to bisect your Spacemacs init file", it would be absolutely wonderful, I think this would help us out a lot.

For context, some details on most recent experience: If someone asks, "I'm having a wonky time with org-capture", and they weren't using Spacemacs, then "bisect" is a pretty easy first suggestion, and empowering. If they are having the same issue with `emacs -Q -f org-capture`, this gets to the point very quickly. These things speed up support, and improve the process, and are very methodical. These approaches are also fairly robust against ill-posed questions, putting less burden on both sides of the question. You don't have to understand the problem perfectly, but if you know what it takes to reproduce the issue, you've gotten quite far in a small amount of time.

**Comment by bradyt** [comment]
I wonder if Spacemacs could have some command similar to -q, so that effectively, it is as if Spacemacs is started with no config.

**Comment by vv111y** [comment]
I'm guilty here.  
Not sure what to do - what is a MWE for spacemacs ??   
Also bisecting: I assume the only meaningful places are in `user-init` and `user-config`   
But what about `org-eval`? I assume that is treated as a whole, and bisect only within.  
Or what if it's some issue between a layer, user-config, and/or upgraded package?
I can't see how Bisecting is straightforward.

**Comment by vv111y** [comment]
Wish we could use this for spacemacs  
https://github.com/Malabarba/elisp-bug-hunter

**Comment by bradyt** [comment]
This seems to help narrow down issues:

> To see if it's your configuration causing the problem, try `mv ~/.spacemacs ~/.spacemacs.bak` and restart emacs.

**Comment by smile13241324** [comment]
I think this is already documented in `Documentation.org#Troubleshooting`:
``` org
** General layer errors
It happens from time to time that some of the layers go stale and stop
working properly for some commands. If this happens stay calm and try
to produce a minimal bug report on the [[https://github.com/syl20bnr/spacemacs/issues][GitHub issue page]].

*** Step by step instructions for minimal layer bug reports
- Load the standard dotfile. You can get it with ~SPC f e D~.   <------
- Add the malfunctioning layer.
- Use ~SPC t D~ to enable debug on error, this will
  produce a backtrace when encountering a lisp error.
- Reproduce the error. Note down the steps and backtrace.
- Press ~SPC h I~ and fill in the place holders
  enclosed in ~<<~ and ~>>~ signs. Don't forget the
  reproduction guide and backtrace.
- Press ~C-c C-c~ or ~, ,~ to transfer your buffers
  content to the github issue creation page.
- Enter a meaningfull title for your issue, stating
  in short words what the issue is.
- When everything has been filled in submit the bug
  report.
- When someone answers to your issue or ask questions
  it goes without saying that you should answer him
  politely and assist in testing a possible fix.
```

**Comment by smile13241324** [comment]
I am not sure if keeping this issue open makes much sense, I have added an extended section about troubleshooting on this subject in the Documentation after this ticket was opened.

Further improvements on this section should be packaged in individual PRs which will be reviewed separately.
